packages/next-server/lib/fetch/retry.ts

```diff
@@ -0,0 +1,77 @@
+export type RetryOptions = {
+  retries?: number
+  factor?: number
+  minTimeout?: number
+  onRetry?(err: Error, attempt: number): void,
+}
+
+export class FetchError extends Error {
+  readonly statusCode: number
+  readonly status: number
+  readonly code: number
+  readonly url: string
+
+  constructor(message: string, { status, url }: { status: number, url: string }) {
+    super(message)
+
+    this.code = this.status = this.statusCode = status
+    this.url = url
+  }
+}
+
+async function sleep(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+export async function retry<R = void>(
+  cb: (bail: (err?: Error) => never, attempt: number, options: RetryOptions) => Promise<R>,
+  options: RetryOptions,
+  attempt: number = 0,
+): Promise<R> {
+  const opts = {
+    // timeouts will be [ 10, 50, 250 ]
+    minTimeout: options.minTimeout || 10,
+    retries: options.retries || 3,
+    factor: options.factor || 5,
+    onRetry: options.onRetry,
+  }
+
+  let aborted = false
+
+  const bail = (err?: Error) => {
+    aborted = true
+    throw err || new Error('Aborted')
+  }
+
+  try {
+    return await cb(bail, attempt, opts)
+  } catch (err) {
+    if (opts.onRetry) opts.onRetry(err, attempt)
+    if (aborted || attempt >= opts.retries) throw err
+
+    await sleep(opts.minTimeout * Math.pow(opts.factor, attempt))
+
+    return retry(cb, opts, attempt + 1)
+  }
+}
+
+export default function fetchRetry(fetch: GlobalFetch['fetch']) {
+  const fn = (url: RequestInfo, opts: RequestInit & { retry?: RetryOptions } = {}) => {
+    return retry(async (_bail, attempt, retryOpts) => {
+      const res = await fetch(url, opts);
+      const isRetry = attempt < (retryOpts.retries || 0)
+
+      // only retry if the request is GET or HEAD
+      if (opts.method && !/(get|head)$/i.test(opts.method)) {
+        return res
+      }
+
+      if (res.status >= 500 && res.status < 600 && isRetry) {
```

Comment:
Imagine the following situation: a backend server gets too high load and starts returning 500 errors, or a proxy in between starts sending 504s. All clients decide to retry their requests on 5XX errors, resulting in even more load on that server, resulting in even more 500s.

Reply:
@Janpot How can you verify that the server is dead and doesn't want more requests to stop retrying ?

Reply:
I don't think you can, I think you just shouldn't retry when the server was able to respond. If the error was recoverable, it should've been retried serverside. Maybe retries could make sense for network-level errors, but IMO, retry logic has so many variants and use-cases that it would probably be better to leave it up to [userland packages](https://www.npmjs.com/package/promise-retry) to implement this.
Anyway, what I really would like to see instead from a polyfill in next.js is the ability to fetch relative urls from server code.

Reply:
@Janpot the native implementation of `fetch` in browsers do it, the retry in this PR tries to do the same for SSR.

Relative URLs is something I want to add too, I just don't know of a reliable way of doing it, I have some ideas about using headers, like `origin`, in the case of `Now` there's a header with the deployment URL but yet it may not always work and I don't know how safe it is.

Reply:
Oh wow, native fetch does this? I wasn't aware, would you mind pointing me to the spec? I find it very hard to believe.

Reply:
@Janpot Mmmm I have no idea, looks like I'm wrong or I'm missing something, I thought it was the default after a conversation with @developit, so take it with a grain of salt 🙏  - In the case that it isn't then I still think is a good default for the reason explained here: https://github.com/zeit/fetch-retry#rationale

Let me know what you think

Reply:
I think, before having non-standard `fetch` extensions, I'd like to see a more portable experience when using it server-side and using it browser-side. Concrete:
- I'd like it to provide a bit more complete polyfill than `unfetch`.
- I'd like it to make relative urls work server-side. Or to polyfill `window.location` server-side. (It shouldn't be based on client provided headers as these can be spoofed easily.)

But ok, on topic, *if* you really want to add retries, you should do it only on network level errors, not on http 5xx responses. An http server should be idempotent. If it was able to respond, I'd expect it to respond in exactly the same way if I retry with the same request, even if it's a 5xx. If a 5xx error is recoverable by retrying, then that retry logic belongs in the server, not in the client.

Also, I'd prefer it to be disabled by default.

That's was my main point. The following are just some secondary thoughts.

>Additionally, if you're adding extensions to the fetch API, another way of eliminating these kinds of errors is by reducing the amount of time the network needs to be used. For instance imaging the following type of code:
>```js
>// /pages/api/my-handler.js
>export default (req, res) => res.send('hello world');
>
>// /pages/my-page/index.js (getInitialState)
>await fetch(import('../api/my-handler'));
>```
>Client-side this could compile to something like
>```js
>await fetch('/api/my-handler');
>```
>while serverside it could import and call that handler directly. Thus eliminating hitting the network. Note that this would also solve the problem of relative urls on the server-side.

Reply:
@Janpot keep in mind that retrying is only for `GET` or `HEAD` requests, maybe we can extend it with better configs, with the downside that Typescript users would have to import `fetch` from `next` so they can have proper types.

The issue with relative imports is basically that we don't know what the absolute URL is, the user may be able to add it with a environment variable, but in that case they can just add the absolute URL by themselves, importing the function code from `/api` may work, the downside is we have to somehow create a request because we can't send the actual request, every api function will always send a response and end the request. This may be worth investigating

Reply:
> keep in mind that retrying is only for GET or HEAD requests

Yep, that's what I meant by "idempotent". It doesn't change my argument though. If the server was able to respond at all, then there is no need to retry, because a well written server will respond with exactly the same status.

> we have to somehow create a request

It's [not so difficult](https://github.com/Janpot/next-serve/blob/0d2a98745c57af430c4f248ec2a0d8eb4af88dc2/src/index.ts#L188) to translate between fetch `Request`/`Response` and node.js `IncomingMessage`/`ServerResponse`. `node-fetch` already does a lot of heavy lifting.